### ISSUE TITLE
kubelet-serial-gce-e2e-hugepages job is flaky

### ISSUE THREAD

**Issue by bart0sh** [kubernetes/kubernetes#99402]
#### Which jobs are flaking:
ci-kubernetes-node-kubelet-serial-hugepages

#### Which test(s) are flaking:
when start the pod with the resources requests that contain multiple hugepages resources should set correct hugetlb mount and limit under the container cgroup

#### Testgrid link:
https://testgrid.k8s.io/sig-node-kubelet#kubelet-serial-gce-e2e-hugepages

#### Reason for failure:
Unexpected error: "pod ran to completion",

#### Anything else we need to know:
Multiple sizes huge pages is going to be graduated to GA, so having this test job fixed is quite important.

**Comment by cynepco3hahue** [comment]
I did a small analysis and I can see that the pod failed to start, I can not find any failure messages under the kubelet log, so it really hard to say why it failed.

```
...
I0224 08:38:40.621] STEP: by running a guarantee pod that requests hugepages
I0224 08:38:40.621] Feb 24 08:38:25.253: INFO: The status of Pod hugepages-hw5qw is Pending, waiting for it to be Running (with Ready = true)
I0224 08:38:40.621] Feb 24 08:38:27.258: INFO: The status of Pod hugepages-hw5qw is Failed which is unexpected
I0224 08:38:40.622] [JustAfterEach] with the resources requests that contain only one hugepages resource 
I0224 08:38:40.622]   _output/local/go/src/k8s.io/kubernetes/test/e2e_node/hugepages_test.go:419
I0224 08:38:40.622] [JustAfterEach] when start the pod
I0224 08:38:40.622]   _output/local/go/src/k8s.io/kubernetes/test/e2e_node/hugepages_test.go:351
I0224 08:38:40.622] STEP: Releasing hugepages
I0224 08:38:40.622] STEP: restarting kubelet to pick up pre-allocated hugepages
I0224 08:38:40.623] Feb 24 08:38:27.279: INFO: Get running kubelet with systemctl: UNIT                            LOAD   ACTIVE SUB     DESCRIPTION                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     
I0224 08:38:40.624] kubelet-20210224T083516.service loaded active running /tmp/node-e2e-20210224T083516/kubelet --kubeconfig /tmp/node-e2e-20210224T083516/kubeconfig --root-dir /var/lib/kubelet --v 4 --logtostderr --feature-gates HugePageStorageMediumSize=true --dynamic-config-dir /tmp/node-e2e-20210224T083516/dynamic-kubelet-config --network-plugin=kubenet --cni-bin-dir /tmp/node-e2e-20210224T083516/cni/bin --cni-conf-dir /tmp/node-e2e-20210224T083516/cni/net.d --cni-cache-dir /tmp/node-e2e-20210224T083516/cni/cache --hostname-override n1-standard-2-ubuntu-gke-1804-1-16-v20200330-032bfc99 --container-runtime docker --container-runtime-endpoint unix:///var/run/dockershim.sock --config /tmp/node-e2e-20210224T083516/kubelet-config --kernel-memcg-notification=true --cgroups-per-qos=true --cgroup-root=/
I0224 08:38:40.625] 
I0224 08:38:40.625] LOAD   = Reflects whether the unit definition was properly loaded.
I0224 08:38:40.625] ACTIVE = The high-level unit activation state, i.e. generalization of SUB.
I0224 08:38:40.625] SUB    = The low-level unit activation state, values depend on unit type.
I0224 08:38:40.625] 
I0224 08:38:40.625] 1 loaded units listed. Pass --all to see loaded but inactive units, too.
I0224 08:38:40.626] To show all installed unit files use 'systemctl list-unit-files'.
I0224 08:38:40.626] , kubelet-20210224T083516
I0224 08:38:40.626] STEP: Waiting for hugepages resource to become available on the local node
I0224 08:38:40.626] I0224 08:38:27.869182    2924 server.go:222] Restarting server "kubelet" with restart command
I0224 08:38:40.626] I0224 08:38:28.028485    2924 server.go:171] Running health check for service "kubelet"
I0224 08:38:40.626] I0224 08:38:28.028517    2924 util.go:48] Running readiness check for service "kubelet"
I0224 08:38:40.627] I0224 08:38:29.029461    2924 server.go:182] Initial health check passed for service "kubelet"
I0224 08:38:40.627] [AfterEach] [sig-node] HugePages [Serial] [Feature:HugePages][NodeSpecialFeature:HugePages]
I0224 08:38:40.627]   /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:186
I0224 08:38:40.627] STEP: Collecting events from namespace "hugepages-test-9184".
I0224 08:38:40.627] STEP: Found 1 events.
I0224 08:38:40.628] Feb 24 08:38:29.385: INFO: At 2021-02-24 08:38:25 +0000 UTC - event for hugepages-hw5qw: {kubelet n1-standard-2-ubuntu-gke-1804-1-16-v20200330-032bfc99} OutOfhugepages-2Mi: Node didn't have enough resource: hugepages-2Mi, requested: 6291456, used: 6291456, capacity: 10485760
I0224 08:38:40.628] Feb 24 08:38:29.393: INFO: POD              NODE                                                   PHASE   GRACE  CONDITIONS
I0224 08:38:40.628] Feb 24 08:38:29.394: INFO: hugepages-hw5qw  n1-standard-2-ubuntu-gke-1804-1-16-v20200330-032bfc99  Failed         []
I0224 08:38:40.629] Feb 24 08:38:29.394: INFO: 
I0224 08:38:40.629] Feb 24 08:38:29.397: INFO: 
I0224 08:38:40.629] Logging node info for node n1-standard-2-ubuntu-gke-1804-1-16-v20200330-032bfc99
...
```

**Comment by bart0sh** [comment]
@cynepco3hahue were you able to reproduce it locally?
btw, there is a "right way" to fix this - to fix dynamic update of a hugepage resources. This would eliminate need in a kubelet restart.  There is even a PR for that: https://github.com/kubernetes/kubernetes/pull/81774, we just need to make it working again.

**Comment by bart0sh** [comment]
@cynepco3hahue after conversation with @derekwaynecarr on slack I'd say we should stick to the current procedure - restarting kubelet when hugepages allocation changes. So, we should investigate why kubelet doesn't restart properly in this test case and fix it.

**Comment by cynepco3hahue** [comment]
Nope, I failed to reproduce it locally, also under the CI, it does not appear often. The best thing will be to describe the pod in case if it failed, so we will see the reason.

**Comment by cynepco3hahue** [comment]
under the code we have
```
setHugepages()

ginkgo.By("restarting kubelet to pick up pre-allocated hugepages")
restartKubelet() <- Feb 28 16:59:49 n1-standard-2-ubuntu-gke-1804-1-16-v20200330-db753712 systemd[1]: Stopping /tmp/node-e2e-20210228T165708/kubelet --kubeconfig /tmp/node-e2e-20210228T165708/kubeconfig --root-dir /var/lib/kubelet --v 4 --logtostderr --feature-gates HugePageStorageMediumSize=true --dynamic-config-dir /tmp/node-e2e-20210228T165708/dynamic-kubelet-config --network-plugin=kubenet --cni-bin-dir /tmp/node-e2e-20210228T165708/cni/bin --cni-conf-dir /tmp/node-e2e-20210228T165708/cni/net.d --cni-cache-dir /tmp/node-e2e-20210228T165708/cni/cache --hostname-override n1-standard-2-ubuntu-gke-1804-1-16-v20200330-db753712 --container-runtime docker --container-runtime-endpoint unix:///var/run/dockershim.sock --config /tmp/node-e2e-20210228T165708/kubelet-config --kernel-memcg-notification=true --cgroups-per-qos=true --cgroup-root=/...

waitForHugepages()

pod := getHugepagesTestPod(f, limits, mounts, volumes)

ginkgo.By("by running a guarantee pod that requests hugepages")
testpod = f.PodClient().CreateSync(pod) <- Feb 28 16:59:49 n1-standard-2-ubuntu-gke-1804-1-16-v20200330-db753712 kubelet[7162]: I0228 16:59:49.623792    7162 config.go:383] Receiving a new pod "hugepages-svgls_hugepages-test-7304(30696202-218c-44e3-9c7b-4cefdc046434)"
```

so from the log https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-node-kubelet-serial-hugepages/1366067610260606976/artifacts/n1-standard-2-ubuntu-gke-1804-1-16-v20200330-db753712/kubelet.log, I can see that the kubelet received the pod before the kubelet stopped, so the problem that pod dropped to Failed state because of the kubelet stop.

**Comment by cynepco3hahue** [comment]
So some additional details, the one who responsible to restart the kubelet again, is the server service:
1. we restart the kubelet service
2. we are waiting for hugepages to be available
3. we are starting the pod
4. the server service restart the kubelet again because it was not healthy for too long - https://github.com/kubernetes/kubernetes/blob/793390e13be91d79150b57abad5710477ab96bd7/test/e2e_node/services/server.go#L222
5. pod failed to start because of the kubelet restart

**Comment by bart0sh** [comment]
@cynepco3hahue great findings! Should we just stop the kubelet and give the server service to start it again? We can also check if the kubelet is restarted successfully before starting the pod.

**Comment by cynepco3hahue** [comment]
@bart0sh I tried it under the PR https://github.com/kubernetes/kubernetes/pull/99584, also I unsure how can we check if the kubelet restarted successfully. All tests that I wrote before I checked if the node has NotReady state and waited until it will be Ready, but in case of the restart probably the node will always have Ready state because the time period is very short. I will continue to investigate and will check locally how can I fix it.

**Comment by cynepco3hahue** [comment]
@bart0sh So the only reliable solution that I found, it just trying to start the pod number of times. I tested it 5 times and the test passed. Can you please review https://github.com/kubernetes/kubernetes/pull/99584?

**Comment by bart0sh** [comment]
@cynepco3hahue Can we access kubelet service from hugepage test cases. If we can we can use its APIs like [ readinessCheck](https://github.com/kubernetes/kubernetes/blob/f79795d71850ab0054b0759f77dd0cec17bb4213/test/e2e_node/services/util.go#L47) or [healthCheck](https://github.com/kubernetes/kubernetes/blob/f79795d71850ab0054b0759f77dd0cec17bb4213/test/e2e_node/services/util.go#L96) to check if kubelet is ready after the restart.

Overall I have a feeling of duplication test/e2e_node/services functionality in test/e23_node/util.go. This is a main reason for this and similar bugs I believe. It would be nice to get rid of this duplication if possible. That would nicely isolate kubelet service controls in one place and eliminate conflicts like this one.

wdyt?

**Comment by cynepco3hahue** [comment]
not really, the `test/e2e_node/services` used mostly to start all needed components to emulate the cluster(not only the kubelet) and the test does not have access to it. I think you can separate the readiness check under the importable method, but it worth the separate PR.

**Comment by cynepco3hahue** [comment]
@bart0sh I updated the PR to check the healthiness of the kubelet and looks like it works just fine.

**Comment by bart0sh** [comment]
@cynepco3hahue thank you. it looks good to me so far. I'll add lgtm label when https://github.com/kubernetes/kubernetes/pull/99634 is merged and tests are all green.

**Comment by bart0sh** [comment]
@cynepco3hahue the job is still flaky. can you reopen this issue?

**Comment by cynepco3hahue** [comment]
@bart0sh I opened a different one - https://github.com/kubernetes/kubernetes/issues/99919, the issue can affect any job, not only the hugepages one.

**Comment by cynepco3hahue** [comment]
I can open a fix to W/A an issue(start the pod until it will succeed), it should fix all problems, WDT?

**Comment by bart0sh** [comment]
i think it's ok as a work around, until #99919 is fixed

**Comment by bart0sh** [comment]
@cynepco3hahue would it be better to wait until api server reports correct resource value *before* running the pod?

**Comment by cynepco3hahue** [comment]
We are waiting for it, the problem that the scheduler and the kubelet see a different picture regarding resources requested by pods.

**Comment by bart0sh** [comment]
ok. In this case restarting the pod until it succeeds seems reasonable to me to fix the flakiness until #99919 is fixed.

**Comment by bart0sh** [comment]
@cynepco3hahue any progress on the workaround fix?
